### PR TITLE
SW-2 | Implement migration analyzer

### PR DESCRIPTION
**Pull request by NikolayProkopchuk**

### CODE REVIEW

File: src/main/java/com/schemawizard/analyzer/impl/PostgresHistoryTableCreator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                    + "id SERIAL PRIMARY KEY, "
+                    + "version INTEGER NOT NULL, "
+                    + "description TEXT NOT NULL, "
+                    + "applied_on TIMESTAMP NOT NULL DEFAULT now(),"
+                    + "checksum INTEGER NOT NULL, "
+                    + "success BOOLEAN NOT NULL DEFAULT FALSE"
+                    + ");";
+    private static final String SELECT_TABLE_SQL = "SELECT table_name " +
+            "FROM information_schema.tables " +
+            "WHERE table_schema='public' AND table_type='BASE TABLE' AND table_name='migration_history';";
```
Comment (AndriyBosik) [inline src/main/java/com/schemawizard/analyzer/impl/PostgresHistoryTableCreator.java:26]: what if application uses schema different from `public`? Or user doesn't have access to `public` schema
Consider inject `ConfigurationProperties` class and use `defaultSchema` property

File: build.gradle [hunk 2]
```diff
@@ -9,9 +9,33 @@ repositories {
...
+    testContainersVersion = '1.19.1'
+    postgresqlVersion = '42.6.0'
+}
+
 dependencies {
-    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.1'
-    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
+    compileOnly "org.projectlombok:lombok:${lombokVersion}"
+
+    implementation "com.google.guava:guava:${guavaVersion}"
```
[the remark this answers is not in the document]
Comment (NikolayProkopchuk) [reply build.gradle:25]: I use this library in ClassesMigrationService in method getMigrations() to get all classes migration, but maybe we can use different approach

File: src/main/java/com/schemawizard/analyzer/impl/ClassesMigrationService.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        try {
+            var classPath = ClassPath.from(ClassLoader.getSystemClassLoader());
+            var topLevelClasses = classPath.getTopLevelClasses(packageName);
+            return topLevelClasses.stream()
+                    .map(ClassPath.ClassInfo::load)
+                    .filter(DbMigration.class::isAssignableFrom)
+                    .map(this::convertMigration)
+                    .sorted(Comparator.comparing(Migration::getVersion))
+                    .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (NikolayProkopchuk) [reply src/main/java/com/schemawizard/analyzer/impl/ClassesMigrationService.java:36]: this service returns all classes migrations, MigrationAnalizer filter not applied and returns them

File: build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndriyBosik) [reply build.gradle:25]: take a look at `Reflections` library as discussed in Slack

**Review by ivlytvynets — CHANGES_REQUESTED** [review]

**Review by ivlytvynets — APPROVED** [review]

File: src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                    + "id SERIAL PRIMARY KEY, "
+                    + "version INTEGER NOT NULL, "
+                    + "description TEXT NOT NULL, "
+                    + "applied_on TIMESTAMP NOT NULL DEFAULT now(),"
+                    + "checksum INTEGER NOT NULL, "
+                    + "success BOOLEAN NOT NULL DEFAULT FALSE"
+                    + ");";
+    private static final String SELECT_TABLE_SQL = "SELECT table_name " +
+            "FROM information_schema.tables " +
+            "WHERE table_type='BASE TABLE' AND table_name='migration_history';";
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java:29]: Would you mind using AbstractFactory pattern like this:
```
public interface HistoryTableQueryFactory {
  String getCreateMigrationHistoryTableSql();
  String getSelectTableSql();
  String getSelectMigrationsSql();
  ...
}

public class PostgresHistoryTableQueryFactory implements HistoryTableQueryFactory {
  // Implementations
}

public class OracleHistoryTableQueryFactory implements HistoryTableQueryFactory {
  // Implementations
}
```
And then, in `SchemaWizardBuilder` we can register in our DI container relevant instance of `HistoryTableQueryFactory` interface based on configuration properties.

And this class might be renamed to `HistoryTableCreatorImpl` with `HistoryTableQueryFactory` as a dependency.

File: src/main/java/com/example/analyzer/service/impl/PostgresAppliedMigrationsService.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresAppliedMigrationsService implements AppliedMigrationsService {
+
+    private static final String SELECT_MIGRATIONS_SQL = "SELECT id, version, description, applied_on " +
+            "FROM migration_history;";
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/service/impl/PostgresAppliedMigrationsService.java:17]: The same as above

File: src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+public class ClassesDeclaredMigrationService implements DeclaredMigrationService {
+
+    private final String packageName;
+
+    private static final Pattern MIGRATION_CLASS_NAME_PATTERN = Pattern.compile("^SW(\\d+)(\\D+\\w*)");
+
+    public ClassesDeclaredMigrationService(String packageName) {
+        this.packageName = packageName;
+    }
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java:24]: Not sure whether our DI container is able to handle String-injection - check it, please
And if it's not able to do this - inject ConfigurationProperties instance

File: src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        String className = migrationClass.getSimpleName();
+        var matcher = MIGRATION_CLASS_NAME_PATTERN.matcher(className);
+        if (!matcher.matches()) {
+            throw new MigrationAnalyzerException(
+                    String.format("Class name '%s' doesn't match the pattern '%s'",
+                            className, MIGRATION_CLASS_NAME_PATTERN.pattern()));
+        }
+
+        int version = Integer.parseInt(matcher.group(1));
+        String description = matcher.group(2);
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java:47]: We also have `@com.example.migration.annotation.SWName` annotation which might be use to declare "custom" migration description. So the correct logic to extract description is:
1. Check if @SWName annotation is present on Migration class
2. If @SWName annotation is present - use description from this annotation
3. If @SWName annotation is absent - use description from Migration classname

File: src/main/java/com/example/dao/ConnectionHolder.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    private Connection connection;
+    private final String url;
+    private final String username;
+    private final String password;
+
+    public ConnectionHolder(String url, String username, String password) {
+        this.url = url;
+        this.username = username;
+        this.password = password;
+    }
```
Comment (AndriyBosik) [inline src/main/java/com/example/dao/ConnectionHolder.java:19]: Not sure whether our DI container is able to handle String injection - check it, please
And if it's not able to do this - inject ConfigurationProperties instance

File: src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            AppliedMigration appliedMigration = appliedMigrations.get(i);
+            DeclaredMigration declaredMigration = declaredMigrations.get(i);
+            compareAppliedAndDeclaredMigrations(appliedMigration, declaredMigration);
+        }
+
+        return declaredMigrations.subList(appliedMigrations.size(), declaredMigrations.size())
+                .stream()
+                .map(this::declaredMigrationToMigrationData)
+                .collect(Collectors.toList());
+    }
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java:49]: In the `AppliedMigrationService.getAppliedMigration()` method, migrations are fetched using this query:
`SELECT id, version, description, applied_on FROM migration_history` -> no ordering was defined and a database might return migrations in any order.
Also, what if a database contains records with versions: 
```
1699126653450
1699126653460
1699126653480
1699126653490
```
And locally user has versions:
```
1699126653450
1699126653460
1699126653470 - new migration
1699126653480
1699126653490
```
This is a possible situation during team work - I can explain it on a call. And does your MigrationAnalyzer is able to handle this case?

File: src/main/java/com/example/analyzer/AppliedMigration.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import java.time.LocalDateTime;
+
+public class AppliedMigration {
+
+    private final Integer id;
+    private final Integer version;
+
+    private final String description;
+
+    private final LocalDateTime appliedOn;
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/AppliedMigration.java:12]: Empty lines are redundant. Remove them from everywhere, please

**Review by AndriyBosik — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java (hunk 6 above)
Comment (NikolayProkopchuk) [reply src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java:24]: ConfigurationProperties.packages is used to specify packages for migrations, so migrations can be located in several packages, or this property is used for another purpose?

File: src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java (hunk 9 above)
Comment (NikolayProkopchuk) [reply src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java:49]: MigrationAnalyzer is not able to handle this case yet, maybe logic should be like revert 1699126653480, 1699126653490 and then return 1699126653470 , 1699126653480, 1699126653490 to execute?

File: src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java (hunk 6 above)
Comment (AndriyBosik) [reply src/main/java/com/example/analyzer/service/impl/ClassesDeclaredMigrationService.java:24]: Currently we don't have any "strict" rules for configuration properties. I defined `packages` as `List` just for example of using lists in yaml file.
So feel free to change it to plain `String` if it's easier for you. But, in this case, do not forget to update `ConfigurationPropertiesServiceImpl.mapConfigurationProperties()` method

File: src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java (hunk 9 above)
Comment (AndriyBosik) [reply src/main/java/com/example/analyzer/impl/MigrationAnalyzerImpl.java:49]: I don't think it's a good solution to revert migration without a user to be asked for. My suggestion is to define `declaredMigrations` like the following:
```
Map<Integer, DeclaredMigration> declaredMigrations = declaredMigrationService.getDeclaredMigrations().stream()
  .collect(
    Collectors.toMap(
      DeclaredMigration::getVersion,
      Function.identity()));
```
And implement comparison:
```
for (AppliedMigration migration: appliedMigrations) {
  if (!declaredMigrations.containsKey(migration.getVersion())) {
    throw new MigrationAnalyzerException("Migration with version " + migration.getVersion() + " was not found");
  }
  declaredMigrations.remove(migration.getVersion());
}
```
Finally, return:
```
return declaredMigrations.values().stream()
  .sorted() // define sorting logic based on migration version
  .collect(Collectors.toList());
```

File: src/main/java/com/example/analyzer/DeclaredMigration.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package com.example.analyzer;
+
+import com.example.migration.Migration;
+
+public class DeclaredMigration implements Comparable<DeclaredMigration> {
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/DeclaredMigration.java:5]: wouldn't it be better to extract this to a separate class or utility class?

File: src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                statement.execute(CREATE_MIGRATION_HISTORY_TABLE_SQL);
+                log.info("Migration history table created.");
+            } else {
+                log.info("Migration history table already exists.");
+            }
+        } catch (SQLException e) {
+            throw new MigrationAnalyzerException("Error creating migration history table: " + e.getMessage(), e);
+        }
+    }
+}
```
Comment (AndriyBosik) [inline src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java:49]: Use PostgresHistoryTableQueryFactory here, please

**Review by AndriyBosik — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java (hunk 12 above)
Comment (NikolayProkopchuk) [reply src/main/java/com/example/analyzer/impl/PostgresHistoryTableCreator.java:49]: sorry, forgot to delete this class, it is not used now)

File: src/main/resources/application.yaml [hunk 13]
```diff
@@ -5,10 +5,7 @@ schema:
       username: postgres
       password: postgres
     migration:
-      packages:
-        - com.example.model
-        - com.example.db.migration
-        - com.example.db.${SOME_ENV_VALUE:default}.downgrade
+      packageName: com.example.db.migration
```
Comment (AndriyBosik) [inline src/main/resources/application.yaml:8]: use `package-name` to follow the same convention

**Review by AndriyBosik — CHANGES_REQUESTED** [review]

**Review by AndriyBosik — APPROVED** [review]

**Review by ivlytvynets — APPROVED** [review]

**Pull request merged**